### PR TITLE
Implement domain layer

### PR DESCRIPTION
**Pull request by vladzhorov**

### CODE REVIEW

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+            order.Price = CalculatePrice(ratePerMinutes, rateDurationMinutes, order.DurationInMinutes);
+
+            await _orderRepository.AddAsync(order, cancellationToken);
+        }
+
+        private decimal CalculatePrice(decimal ratePerMinutes, int rateDurationMinutes, int durationInMinutes)
+        {
+            decimal ratePerMinute = ratePerMinutes / rateDurationMinutes;
+            return Math.Round(ratePerMinute * durationInMinutes, 2);
```
Comment (moltenessence) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs:47]: What's the difference between ratePerMinutes and ratePerMinute? rateDurationMinutes vs durationInMinutes?

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            order.Status = OrderStatus.Pending;
+
+            order.Price = CalculatePrice(ratePerMinutes, rateDurationMinutes, order.DurationInMinutes);
+
+            await _orderRepository.AddAsync(order, cancellationToken);
+        }
+
+        private decimal CalculatePrice(decimal ratePerMinutes, int rateDurationMinutes, int durationInMinutes)
+        {
+            decimal ratePerMinute = ratePerMinutes / rateDurationMinutes;
```
Comment (moltenessence) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs:46]: decimal -> var

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Validators/OrderStatusValidator.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+{
+    public static class OrderStatusValidator
+    {
+        private static readonly Dictionary<OrderStatus, OrderStatus[]> ValidOrderStatusTransitions = new Dictionary<OrderStatus, OrderStatus[]>
+        {
+            { OrderStatus.Pending, new OrderStatus[] { OrderStatus.Confirmed, OrderStatus.Cancelled } },
+            { OrderStatus.Confirmed, new OrderStatus[] { OrderStatus.InProgress, OrderStatus.Cancelled } },
+            { OrderStatus.InProgress, new OrderStatus[] { OrderStatus.Completed, OrderStatus.Cancelled } },
+        };
+
```
Comment (moltenessence) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Validators/OrderStatusValidator.cs:13]: dictionary can be immutable

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Validators/OrderStatusValidator.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        private static readonly Dictionary<OrderRequestStatus, OrderRequestStatus[]> ValidOrderRequestStatusTransitions = new Dictionary<OrderRequestStatus, OrderRequestStatus[]>
+        {
+            { OrderRequestStatus.Pending, new OrderRequestStatus[] { OrderRequestStatus.Approved, OrderRequestStatus.Rejected } },
+            { OrderRequestStatus.Approved, new OrderRequestStatus[] { OrderRequestStatus.Cancelled } },
+        };
+
```
Comment (moltenessence) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Validators/OrderStatusValidator.cs:19]: same

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderRequestService.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            OrderStatusValidator.ValidateStatusTransition(orderRequest.Status, newStatus);
+
+            orderRequest.Status = newStatus;
+            orderRequest.UpdatedAt = DateTime.UtcNow;
+
+            await _orderRequestRepository.UpdateAsync(orderRequest, cancellationToken);
+            _domainEventDispatcher.Dispatch(new OrderRequestStatusChangedEvent(orderRequest.Id, newStatus));
+        }
+
+
```
Comment (moltenessence) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderRequestService.cs:34]: formatting

**Review by moltenessence — CHANGES_REQUESTED** [review]

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs (hunk 1 above)
Comment (Mmishaaa) [reply ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs]: in spelling

Comment (vladzhorov) [reply ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs]: ))

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderRequestService.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        public async Task UpdateOrderRequestStatusAsync(Guid orderRequestId, OrderRequestStatus newStatus, CancellationToken cancellationToken)
+        {
+            var orderRequest = await _orderRequestRepository.GetByIdAsync(orderRequestId, cancellationToken);
+
+            OrderStatusValidator.ValidateStatusTransition(orderRequest.Status, newStatus);
+
+            orderRequest.Status = newStatus;
+            orderRequest.UpdatedAt = DateTime.UtcNow;
```
Comment (daniilbluetape) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderRequestService.cs:28]: implement date provider

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderRequestService.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+            await _orderRequestRepository.UpdateAsync(orderRequest, cancellationToken);
+            _domainEventDispatcher.Dispatch(new OrderRequestStatusChangedEvent(orderRequest.Id, newStatus));
+        }
+
+        public async Task CreateOrderRequestAsync(OrderRequest orderRequest, CancellationToken cancellationToken)
+        {
+            orderRequest.CreatedAt = DateTime.UtcNow;
+            orderRequest.UpdatedAt = DateTime.UtcNow;
```
Comment (daniilbluetape) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderRequestService.cs:37]: implement date provider

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        public async Task UpdateOrderStatusAsync(Guid orderId, OrderStatus newStatus, CancellationToken cancellationToken)
+        {
+            var order = await _orderRepository.GetByIdAsync(orderId, cancellationToken);
+
+            OrderStatusValidator.ValidateStatusTransition(order.Status, newStatus);
+
+            order.Status = newStatus;
+            order.UpdatedAt = DateTime.UtcNow;
```
Comment (daniilbluetape) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Services/OrderService.cs:27]: implement date provider

File: ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Validators/OrderStatusValidator.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        private static readonly ImmutableDictionary<OrderStatus, ImmutableArray<OrderStatus>> ValidOrderStatusTransitions =
+                 ImmutableDictionary<OrderStatus, ImmutableArray<OrderStatus>>.Empty
+                .Add(OrderStatus.Pending, [OrderStatus.Confirmed, OrderStatus.Cancelled])
+                .Add(OrderStatus.Confirmed, [OrderStatus.InProgress, OrderStatus.Cancelled])
+                .Add(OrderStatus.InProgress, [OrderStatus.Completed, OrderStatus.Cancelled]);
+
+        private static readonly ImmutableDictionary<OrderRequestStatus, ImmutableArray<OrderRequestStatus>> ValidOrderRequestStatusTransitions =
+            ImmutableDictionary<OrderRequestStatus, ImmutableArray<OrderRequestStatus>>.Empty
+                .Add(OrderRequestStatus.Pending, [OrderRequestStatus.Approved, OrderRequestStatus.Rejected])
+                .Add(OrderRequestStatus.Approved, [OrderRequestStatus.Cancelled]);
```
Comment (daniilbluetape) [inline ServiceFinder.OrderService/ServiceFinder.OrderService.Domain/Validators/OrderStatusValidator.cs:17]: You could use .ToImmutableDictionary extension

**Review by daniilbluetape — CHANGES_REQUESTED** [review]

**Review by moltenessence — APPROVED** [review]

**Review by moltenessence — COMMENTED** [review]
delete

**Pull request merged**